### DISCUSSION TITLE
Migrating to stdout and stderr write from print

### DISCUSSION THREAD

**Discussion by yahu1031** [serverpod/serverpod#138] [Ideas]
Instead of using print statements, How about moving to `stdout.write` and `stderr.write`? This will make much sense.

**Comment by MarkOSullivan94** [comment]
A couple other alternatives which could be used.

https://pub.dev/packages/logging

https://pub.dev/packages/classic_logger

**Reply by yahu1031** [reply]
Yes, @MarkOSullivan94 they are good but for debugging. But I'm talking about printing in console on production mode. Cause print statements are not good on production mode. I think they are in flutter. But we have a standard option of using it.
CC: @vlidholt

**Comment by vlidholt** [comment]
That is done for most parts for internal logging within Serverpod itself. For logging within your code you typically want to use the `session.log()` method which provides really nice logging and a visual viewer (soon!).

**Reply by yahu1031** [reply]
@vlidholt Sorry, I missed this update and forgot about this. But yeah, I'm not mentioning logs, but the print wraps we have in the CLI. To follow the standard syntax for CLIs, We better consider using `std` protocols. When I'm trying to communicate with CLI from the outer source, I can't see the `Docker not started/installed` error as an error. It prints as stdout with exit code `0` instead of stderr with `ErrorCode`. Any ideas about it? One of my whole projects depending on the CLI was held back.

**Reply by yahu1031** [reply]
@vlidholt - remainder please

**Comment by SandPod** [comment]
@yahu1031 We did a bigger overhaul of the logging from the CLI available in main now.

Does this solve the issue you ran into?

The CLI now uses stderr and stdout depending on the log level of the logged message.
https://github.com/serverpod/serverpod/blob/main/tools/serverpod_cli/lib/src/logger/loggers/std_out_logger.dart
